File: python_kata/src/trees/binary_tree/in_order_traversal.py
from typing import List
from trees.binary_tree.node import BinaryNode


def walk(curr: BinaryNode[int] | None, path: List[int]) -> List[int]:
    if not curr:
        return path

    # Pre

    # Recursive
    walk(curr.left, path)
    path.append(curr.val)
    walk(curr.right, path)

    # Post
    return path

def in_order_traversal(head: BinaryNode[int]) -> List[int]:
    return walk(head, [])

File: typescript-kata/src/trees/binary-tree/inOrderTraversal.ts
import { BinaryNode, BinaryTree } from "./node";

function walk(curr: BinaryNode<number> | undefined, path: number[]): number[] {
  if (!curr) {
    return path;
  }

  // Pre
  //
  // Recursive
  walk(curr.left, path);
  path.push(curr.val);
  walk(curr.right, path);

  // Post
  return path;
}
export default function InOrderTraversal(tree: BinaryTree<number>): number[] {
  if (!tree.root) {
    return [];
  }
  return walk(tree.root, []);
}
